openhcl/virt_mshv_vtl/src/processor/hardware_cvm/mod.rs

```diff
@@ -380,20 +369,112 @@ impl<T, B: HardwareIsolatedBacking> hv1_hypercall::SetVpRegisters
             .map_err(|_| (HvError::InvalidParameter, 0))?;
 
         for (i, reg) in registers.iter().enumerate() {
-            if reg.name == hvdef::HvX64RegisterName::VsmPartitionConfig.into() {
-                let value = HvRegisterVsmPartitionConfig::from(reg.value.as_u64());
-                self.vp
-                    .set_vsm_partition_config(value, target_vtl)
-                    .map_err(|e| (e, i))?;
-            } else {
-                return Err((HvError::InvalidParameter, i));
+            match HvX64RegisterName::from(reg.name) {
+                HvX64RegisterName::VsmPartitionConfig => {
+                    self.vp
+                        .set_vsm_partition_config(
+                            HvRegisterVsmPartitionConfig::from(reg.value.as_u64()),
+                            target_vtl,
+                        )
+                        .map_err(|e| (e, i))?;
+                }
+                HvX64RegisterName::VpAssistPage => {
+                    self.vp.hv[target_vtl]
+                        .as_mut()
+                        .expect("has hv emulator")
+                        .msr_write(hvdef::HV_X64_MSR_VP_ASSIST_PAGE, reg.value.as_u64())
+                        .map_err(|_| (HvError::InvalidRegisterValue, 0))?;
+                }
+                _ => {
+                    tracing::error!(
+                        ?reg,
+                        "guest invoked SetVpRegisters with unsupported register"
+                    );
+                    return Err((HvError::InvalidParameter, i));
+                }
             }
         }
 
         Ok(())
     }
 }
 
+impl<T, B: HardwareIsolatedBacking> hv1_hypercall::VtlCall for UhHypercallHandler<'_, '_, T, B> {
+    fn is_vtl_call_allowed(&self) -> bool {
+        tracing::trace!("checking if vtl call is allowed");
+
+        // Only allowed from VTL 0
+        if self.intercepted_vtl != GuestVtl::Vtl0 {
+            false
+        } else if !*self.vp.inner.hcvm_vtl1_enabled.lock() {
+            // VTL 1 must be enabled on the vp
+            false
+        } else {
+            true
+        }
+    }
+
+    fn vtl_call(&mut self) {
+        tracing::trace!("handling vtl call");
+
+        B::switch_vtl_state(self.vp, self.intercepted_vtl, GuestVtl::Vtl1);
+        self.vp.backing.cvm_state_mut().exit_vtl = GuestVtl::Vtl1;
```

Comment:
There's no argument to set the target VTL, correct? It's implied that VtlCall from VTL0 always goes to VTL1? (Not that there's anywhere else to go.)

Reply:
VtlCall can only go into the next highest VTL, and we check in is_vtl_call_allowed that the intercepted VTL was VTL 0, so we can only go 0 -> 1.